internal/zaputil/logrus.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2021, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package zaputil
+
+import (
+	"fmt"
+	"github.com/sirupsen/logrus"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"io/ioutil"
+	"sync"
+)
+
+type zapWrapper zap.Logger
+
+type redirect struct {
+	from *logrus.Logger
+	to *zap.Logger
+}
+
+var (
+	mu        sync.Mutex
+	redirects = make(map[redirect]bool)
+	levelsMap = newLevelsMap()
+)
+
+// RedirectLogrusLogs mutes the output of the supplied logrus.Logger and adds a hook to it.
+// When fired the hook creates a zap entry using the supplied logrus entry.
+func RedirectLogrusLogs(from *logrus.Logger, to *zap.Logger) {
+	mu.Lock()
+	defer mu.Unlock()
+	r := redirect{from, to}
+	if redirects[r] {
+		return
+	}
+	from.ReportCaller = true
+	from.SetOutput(ioutil.Discard)
+	from.AddHook((*zapWrapper)(to))
```

Comment:
`from` is always logrus's exported std logger right?  This would mean that all agent logs would be redirected to all smart agent receiver loggers.  Are we sure there is only one central zap logger that all components share?

Reply:
appears each receiver gets its own: https://github.com/open-telemetry/opentelemetry-collector/blob/1d38caf3b757b4d2f3db10ce9dc04187f170705f/service/builder/receivers_builder.go#L106

Reply:
Yes, there's no guarantee that all components share a single zap logger and we could run into duplicate log entries. We could new up a zap logger and redirect to it, I think.